File: ios/LaneShadow/Views/Molecules/DateRangePicker.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Date Range Preset

/**
 * Date range preset options
 *
 * Defines the available preset time ranges for filtering.
 */
public enum LSDateRangePreset: String, CaseIterable, Sendable {
    case all = "All time"
    case week = "Last week"
    case month = "Last month"
    case threeMonths = "Last 3 months"

    /// Returns the number of days to go back from today (nil for "all time")
    var daysBack: Int? {
        switch self {
        case .all: nil
        case .week: 7
        case .month: 30
        case .threeMonths: 90
        }
    }
}

// MARK: - Date Range Model

/**
 * Date range model
 *
 * Represents a filtered date range with optional bounds.
 */
public struct LSDateRange: Sendable, Equatable {
    public var afterDate: Date?
    public var beforeDate: Date?

    public init(afterDate: Date? = nil, beforeDate: Date? = nil) {
        self.afterDate = afterDate
        self.beforeDate = beforeDate
    }

    /// Empty date range (no filtering)
    public static let empty = LSDateRange()
}

// MARK: - Date Range Picker Component

/**
 * Date range picker molecule component
 *
 * Chip-style date range selector with preset options for filtering by creation date.
 * Following React Native component from react-native/components/ui/date-range-picker.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Selected background: `theme.colors.primary.default`
 *   - Selected text: `theme.colors.onPrimary.default`
 *   - Unselected background: `theme.colors.surfaceVariant.default`
 *   - Unselected text: `theme.colors.onSurface.default`
 *   - Border: `theme.colors.border.default` (1pt)
 * - Layout:
 *   - Corner radius: pill shape (`theme.radius.full`)
 *   - Padding: horizontal 12pt, vertical 6pt
 *   - Spacing between chips: 6pt
 *   - Horizontal scroll padding: 12pt
 * - Typography:
 *   - Label: `theme.type.label.sm` (small, semibold)
 *
 * ## Behavior
 * - Default selected: `.all`
 * - Tapping same chip toggles back to `.all`
 * - On selection, computes `afterDate = now - daysBack` days
 * - Horizontal scrolling when chips overflow
 *
 * ## Parameters
 * - onDateRangeChange: Callback when range changes with computed `LSDateRange`
 * - testID: Optional testing identifier for UI tests
 */
public struct LSDateRangePicker: View {
    @Environment(\.theme) private var theme
    @State private var selectedPreset: LSDateRangePreset = .all

    private let onDateRangeChange: (LSDateRange) -> Void
    private let testID: String

    public init(
        onDateRangeChange: @escaping (LSDateRange) -> Void,
        testID: String = "date-range-picker"
    ) {
        self.onDateRangeChange = onDateRangeChange
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: theme.space.sm) {
                ForEach(LSDateRangePreset.allCases, id: \.self) { preset in
                    chipButton(for: preset)
                }
            }
            .padding(.horizontal, theme.space.md)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Date range filter")
    }

    // MARK: - Chip Button

    @ViewBuilder
    private func chipButton(for preset: LSDateRangePreset) -> some View {
        let isSelected = selectedPreset == preset

        Button(action: {
            handlePresetTap(preset)
        }) {
            Text(preset.rawValue)
                .font(.system(size: theme.type.label.sm.fontSize, weight: theme.type.label.sm.fontWeight))
                .foregroundStyle(isSelected ? theme.colors.onPrimary.default : theme.colors.onSurface.default)
                .padding(.horizontal, theme.space.md)
                .padding(.vertical, theme.space.sm)
                .background(
                    RoundedRectangle(cornerRadius: theme.radius.full)
                        .fill(isSelected ? theme.colors.primary.default : theme.colors.surfaceVariant.default)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: theme.radius.full)
                        .stroke(theme.colors.border.default, lineWidth: isSelected ? 0 : 1)
                )
        }
        .buttonStyle(PlainButtonStyle())
        .accessibilityLabel(preset.rawValue)
        .accessibilityValue(isSelected ? "Selected" : "Not selected")
        .accessibilityAddTraits(isSelected ? [.isSelected] : [])
        .accessibilityIdentifier(
            "\(testID)-chip-\(preset.rawValue.lowercased().replacingOccurrences(of: " ", with: "-"))"
        )
    }

    // MARK: - Interaction Handlers

    private func handlePresetTap(_ preset: LSDateRangePreset) {
        if preset == .all {
            selectedPreset = .all
            onDateRangeChange(.empty)
            return
        }

        if preset == selectedPreset {
            // Toggle off — deselect back to "All time"
            selectedPreset = .all
            onDateRangeChange(.empty)
            return
        }

        selectedPreset = preset
        onDateRangeChange(computeDateRange(for: preset))
    }

    private func computeDateRange(for preset: LSDateRangePreset) -> LSDateRange {
        guard let daysBack = preset.daysBack else {
            return .empty
        }

        let calendar = Calendar.current
        let now = Date()
        guard let afterDate = calendar.date(byAdding: .day, value: -daysBack, to: now) else {
            return .empty
        }

        return LSDateRange(afterDate: afterDate, beforeDate: nil)
    }
}

// MARK: - Preview

#Preview("DateRangePicker - Default") {
    LSDateRangePicker { range in
        print("Date range changed:")
        print("  After: \(range.afterDate?.description ?? "none")")
        print("  Before: \(range.beforeDate?.description ?? "none")")
    }
    .laneShadowTheme()
}

#Preview("DateRangePicker - With Initial Selection") {
    VStack(spacing: 20) {
        Text("Filter rides by date")
            .font(.headline)

        LSDateRangePicker { range in
            print("Date range changed:")
            print("  After: \(range.afterDate?.description ?? "none")")
            print("  Before: \(range.beforeDate?.description ?? "none")")
        }

        Text("Selected range will appear above")
            .font(.caption)
            .foregroundStyle(.secondary)
    }
    .laneShadowTheme()
}

#Preview("DateRangePicker - Light Mode") {
    LSDateRangePicker { range in
        print("Range: \(range.afterDate?.description ?? "none")")
    }
    .laneShadowTheme()
    .preferredColorScheme(.light)
}

#Preview("DateRangePicker - Dark Mode") {
    LSDateRangePicker { range in
        print("Range: \(range.afterDate?.description ?? "none")")
    }
    .laneShadowTheme()
    .preferredColorScheme(.dark)
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/DateRangePicker.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Date range preset options
 *
 * Following RN wrapper API from react-native/components/ui/date-range-picker.tsx
 */
enum class DateRangePreset(val label: String, val daysBack: Int?) {
    ALL("All time", null),
    WEEK("Last week", 7),
    MONTH("Last month", 30),
    THREE_MONTHS("Last 3 months", 90),
}

/**
 * Date range data type
 *
 * Following RN wrapper API from react-native/components/ui/date-range-picker.tsx
 */
data class DateRange(
    val afterDate: Long? = null,
    val beforeDate: Long? = null,
)

/**
 * DateRangePicker molecule component
 *
 * Chip-style date range picker with preset options for filtering routes by creation date.
 * Following React Native wrapper patterns from react-native/components/ui/date-range-picker.tsx
 *
 * @param onDateRangeChange Callback when date range changes
 * @param modifier Modifier for the picker container
 * @param testId Test ID for UI testing
 */
@Composable
fun DateRangePicker(
    onDateRangeChange: (DateRange) -> Unit,
    modifier: Modifier = Modifier,
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // State for selected preset (default: ALL)
    var selected by rememberSaveable { mutableStateOf(DateRangePreset.ALL) }

    // Compute date range for a preset
    fun getDateRange(preset: DateRangePreset): DateRange {
        val daysBack = preset.daysBack ?: return DateRange(afterDate = null, beforeDate = null)
        val afterDate = System.currentTimeMillis() - (daysBack * 24 * 60 * 60 * 1000L)
        return DateRange(afterDate = afterDate, beforeDate = null)
    }

    // Handle chip press
    fun handlePress(preset: DateRangePreset) {
        if (preset == DateRangePreset.ALL) {
            selected = DateRangePreset.ALL
            onDateRangeChange(DateRange(afterDate = null, beforeDate = null))
            return
        }
        if (preset == selected) {
            // Toggle off — deselect back to "All time"
            selected = DateRangePreset.ALL
            onDateRangeChange(DateRange(afterDate = null, beforeDate = null))
            return
        }
        selected = preset
        onDateRangeChange(getDateRange(preset))
    }

    // Build semantics
    val pickerModifier = modifier.semantics {
        contentDescription = testId ?: "date-range-picker"
    }

    Surface(
        modifier = pickerModifier,
        color = theme.colors.background.default,
    ) {
        Row(
            modifier = Modifier
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = theme.space.md),
        ) {
            DateRangePreset.entries.forEachIndexed { index, preset ->
                if (index > 0) {
                    // Spacing between chips
                    androidx.compose.foundation.layout.Spacer(modifier = Modifier.width(theme.space.sm))
                }

                val isSelected = selected == preset

                // Chip background color
                val chipBackgroundColor = if (isSelected) {
                    theme.colors.primary.default
                } else {
                    theme.colors.surfaceVariant.default
                }

                // Chip text color
                val chipTextColor = if (isSelected) {
                    theme.colors.onPrimary.default
                } else {
                    theme.colors.onSurface.default
                }

                // Chip border (only for unselected)
                val chipBorder = if (!isSelected) {
                    BorderStroke(1.dp, theme.colors.border.default)
                } else {
                    null
                }

                val chipTestId = "${testId ?: "date-range-picker"}-chip-${preset.name.lowercase()}"

                Surface(
                    color = chipBackgroundColor,
                    shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.full),
                    border = chipBorder,
                    modifier = Modifier
                        .clickable { handlePress(preset) }
                        .semantics {
                            contentDescription = chipTestId
                        }
                ) {
                    Text(
                        text = preset.label,
                        style = theme.type.label.sm,
                        color = chipTextColor,
                        modifier = Modifier.padding(
                            horizontal = theme.space.md,
                            vertical = theme.space.sm,
                        ),
                    )
                }
            }
        }
    }
}
